x-pack/plugins/gis/public/angular/workspace_controller.js

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import chrome from 'ui/chrome';
+import React from 'react';
+import { render, unmountComponentAtNode } from 'react-dom';
+import { uiModules } from 'ui/modules';
+import { applyTheme } from 'ui/theme';
+import { timefilter } from 'ui/timefilter';
+import { Provider } from 'react-redux';
+import { getStore } from '../store/store';
+import { GISApp } from '../components/gis_app';
+import { setTimeFilters, mapExtentChanged, replaceLayerList } from '../actions/store_actions';
+import { getIsDarkTheme } from '../store/ui';
+import { Inspector } from 'ui/inspector';
+import { inspectorAdapters } from '../kibana_services';
+import { SavedObjectSaveModal } from 'ui/saved_objects/components/saved_object_save_modal';
+import { showSaveModal } from 'ui/saved_objects/show_saved_object_save_modal';
+import { showOptionsPopover } from '../components/top_nav/show_options_popover';
+import { toastNotifications } from 'ui/notify';
+import { getMapReady } from "../selectors/map_selectors";
+
+const REACT_ANCHOR_DOM_ELEMENT_ID = 'react-gis-root';
+
+const app = uiModules.get('app/gis', []);
+
+app.controller('GisWorkspaceController', ($scope, $route, config, breadcrumbState, kbnUrl) => {
+
+  let isLayersListInitializedFromSavedObject = false;
+  const savedWorkspace = $scope.workspace = $route.current.locals.workspace;
+  let isDarkTheme;
+  let unsubscribe;
+
+  getStore().then(store => {
+    handleStoreChanges(store);
```

Comment:
It looks like you're watching the store for changes, but the store object changes pretty frequently. Could the propagated changes be fired using selectors which only fire when their specific return values change?